public/app/plugins/datasource/influxdb/components/queryUtils.ts

```diff
@@ -0,0 +1,57 @@
+import { cloneDeep } from 'lodash';
+import InfluxQueryModel from '../influx_query_model';
+import { InfluxQuery } from '../types';
+
+// FIXME: these functions are a beginning of a refactoring of influx_query_model.ts
+// into a simpler approach with full typescript types.
+// later we should be able to migrate the unit-tests
+// that relate to these functions here, and then perhaps even move the implementation
+// to this place
+
+export function buildRawQuery(query: InfluxQuery): string {
+  const queryCopy = { ...query }; // the query-model mutates the query
+  const model = new InfluxQueryModel(queryCopy);
+  return model.render(false);
+}
+
+export function normalizeQuery(query: InfluxQuery): InfluxQuery {
+  const queryCopy = { ...query }; // the query-model mutates the query
```

Comment:
same as before, maybe you want to use `cloneDeep` here as well?

Reply:
done. i also added some code to check if changes are needed or not, and if not, the original query is returned. i need this  because otherwise the (normalized) query-object changes in every render, and `useMemo` will not be efficient.

Reply:
Cool! 🚀